### ISSUE TITLE
carbonapi fails to return result when format=pickle

### ISSUE THREAD

**Issue by sbskas** [go-graphite/carbonapi#379]
We have problems to make graphite-web frontend query apiserver.

After having setup go-carbon server and apiserver on front of it, we struggled to make the graphite api display metrics.

Result query returns empty.

Here some example :
`
curl 'http://localhost:8080/metrics/find/?format=treejson&query=%2A'
[{"allowChildren":1,"expandable":1,"leaf":0,"id":"app1","text":"carbon","context":{}},{"allowChildren":1,"expandable":1,"leaf":0,"id":"app2","text":"poubelle","context":{}},{"allowChildren":1,"expandable":1,"leaf":0,"id":"app3","text":"valid","context":{}}
` 
Whereas the same query in pickle format returns 0 result:

>curl -v 'http://localhost:8080/metrics/find/?format=pickle&query=%2A'
>* About to connect() to localhost port 8080 (#0)
>*   Trying 127.0.0.1...
>* Connected to localhost (127.0.0.1) port 8080 (#0)
>> GET /metrics/find/?format=pickle&query=%2A HTTP/1.1
>> User-Agent: curl/7.29.0
>> Host: localhost:8080
>> Accept: */*
>> 
>< HTTP/1.1 200 OK
>< Content-Type: application/pickle
>< Date: Wed, 28 Nov 2018 17:35:30 GMT
>< Content-Length: 0
>< 
>* Connection #0 to host localhost left intact

Content-length is 0.

Testing with grafana works ok.

**Comment by dgryski** [comment]
I wonder if we're returning an error from the encoder and not logging it.

**Comment by sbskas** [comment]
Maybe, this completly broke our graphite frontend.
I also made some test with other backend.
The only one that is reliable is protobuf.
All others seems to be broken (even msgpack).

**Comment by azhiltsov** [comment]
I might be wrong here, but graphite-web should be talking to carbonzipper, not carbonapi. 
It uses CLUSTER_SERVERS config parameter in order to configure the endpoint.

**Comment by sbskas** [comment]
Ok I'm now officially puzzled.
There was a big notice in the main README saying : 

> **CarbonAPI should now do everything Carbonzipper used to. It's probably the project you're interested in.**

My understanding of this sentence is that apiserver is now a drop-in replacement for carbonzipper.
May I have some explanations ?

**Comment by Civil** [comment]
carbonzipper git repo was merged into carbonapi one. You can build both tools from the same repo. And indeed there are plans to deprecated zipper cmd.

**Comment by reyjrar** [comment]
# Problem

Neither `carbonapi` nor `carbonzipper` work as a `CLUSTER_SERVERS` member in graphite-web 1.1.4. 

# Setup

I'm running:

carbonapi (built from master as of 2019-04-22)
carbonzipper (built from master as of 2019-04-22)
go-carbon 0.13
graphite-web 1.1.4

I have 2, 3 node clusters of storage nodes running go-carbon with carbonserver enabled with:

```ini
[carbonserver]
listen = ":8080"
enabled = true
buckets = 10
metrics-as-counters = false
read-timeout = "60s"
write-timeout = "60s"
query-cache-enabled = true
query-cache-size-mb = 8096
find-cache-enabled = true
trigram-index = true
scan-frequency = "5m0s"
max-globs = 100
fail-on-max-globs = false
graphite-web-10-strict-mode = true
internal-stats-dir = ""
stats-percentiles = [99, 95, 75, 50]
```
I cannot seem to get either carbonzipper nor carbonapi to work as a CLUSTER_SERVER for graphite-web 1.1.4.

If I try direct to a group a carbonserver nodes with carbonapi settings:

```yaml
upstreams:
    backendsv2:
        backends:
          -
             groupName: "misc"
             protocol: "auto"
             lbMethod: "broadcast"
             servers:
                - "http://graphite3a:8080"  # carbonserver from go-carbon 0.13
                - "http://graphite5a:8080"  # carbonserver from go-carbon 0.13
                - "http://graphite7a:8080"  # carbonserver from go-carbon 0.13
```
Everything in graphite-web 1.1.4 works **except**  no autocomplete, no metrics tree browser and requests to `/metrics/find/` 500 in graphite-web.

## Logs from graphite-web 1.1.4

Attempting to load the metrics tree in graphite-web:

```
2019-04-23,16:14:30.426 :: RemoteFinder[127.0.0.1:8082] Error decoding find response from http://127.0.0.1:8082/metrics/find/?local=1&format=pickle&query=%2A:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/graphite/finders/remote.py", line 115, in find_nodes
    result, buffer_size=settings.REMOTE_BUFFER_SIZE))
  File "/usr/local/lib/python2.7/site-packages/graphite/util.py", line 186, in load
    return pickle_obj.load()
EOFError
2019-04-23,16:14:30.427 :: Exception Caught
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/graphite/metrics/views.py", line 128, in find_view
    leaves_only=leaves_only,
  File "/usr/local/lib/python2.7/site-packages/graphite/storage.py", line 264, in find
    for match in self._find(query):
  File "/usr/local/lib/python2.7/site-packages/graphite/storage.py", line 295, in _find
    for node in result or []:
  File "/usr/local/lib/python2.7/site-packages/graphite/finders/remote.py", line 121, in find_nodes
    raise Exception("Error decoding find response from %s: %s" % (result.url_full, err))
Exception: Error decoding find response from http://127.0.0.1:8082/metrics/find/?local=1&format=pickle&query=%2A:
2019-04-23,16:14:30.428 :: Exception encountered in <GET http://graphiteweb2a.int.craigslist.org/metrics/find/?_dc=1556061270315&query=*&format=treejson&path=&node=GraphiteTree>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python2.7/site-packages/graphite/metrics/views.py", line 128, in find_view
    leaves_only=leaves_only,
  File "/usr/local/lib/python2.7/site-packages/graphite/storage.py", line 264, in find
    for match in self._find(query):
  File "/usr/local/lib/python2.7/site-packages/graphite/storage.py", line 295, in _find
    for node in result or []:
  File "/usr/local/lib/python2.7/site-packages/graphite/finders/remote.py", line 121, in find_nodes
    raise Exception("Error decoding find response from %s: %s" % (result.url_full, err))
Exception: Error decoding find response from http://127.0.0.1:8082/metrics/find/?local=1&format=pickle&query=%2A:
```

# carbonzipper by itself

If I just use carbonzipper via the same backend config, I get autocomplete and the metric tree browser working as expected.  I can even fetch fully qualified metrics, but **any** fetch with a wildcard results in a zero byte, 200 response.  Any request with a function also similarly returns nothing.

# Using both carobnzipper and carbonapi

So I thought, what if I put them together?  OK, with carbonzipper running the above config, listening on :8080 and carbonapi configured to use the localhost:8080 as it's only backend and listening on :8082, when I plop localhost:8082 into the `CLUSTER_SERVERS` for graphite-web 1.1.4 it behaves identically to the initial configuration.  Queries work and return the expected data but the metrics tree browser and autocomplete are not working, resulting in the same log as previously shown.

# Result

`carbonzipper` appears to do the right thing with `pickle` queries from graphite-web but can't handle queries that have wildcards or functions.  `carbonapi` works fine, but calls in `/metrics/` API land appear to return invalid `pickle` responses for graphite-web.

Please send help. ;)

**Comment by reyjrar** [comment]
This might be related to https://github.com/graphite-project/graphite-web/pull/2354 which I'm going to build and test against tomorrow.

**Comment by reyjrar** [comment]
So just upgrading didn't work, though the fail is coming from a different part of graphite-web now:

```
2019-04-24,09:59:02.915 :: RemoteFinder[127.0.0.1:8082] Error decoding response from http://127.0.0.1:8082/metrics/find/?local=1&format=pickle&query=%2A:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/graphite/finders/remote.py", line 291, in deserialize
    return self._deserialize_stream(reader, result.getheader('content-type'))
  File "/usr/local/lib/python2.7/site-packages/graphite/finders/remote.py", line 319, in _deserialize_stream
    data = unpickle.load(stream)
  File "/usr/local/lib/python2.7/site-packages/graphite/util.py", line 186, in load
    return pickle_obj.load()
EOFError
```

**Comment by reyjrar** [comment]
And yes, carbonapi is returning HTTP 200, size 0 bytes responses for `/metrics/find/?format=pickle` which is broken.

**Comment by reyjrar** [comment]
After upgrading to graphiteweb-1.1.5 and using carbonzipper as the `CLUSTER_SERVERS` member for graphite-web, I am able to view the metrics tree, but requests with *wildcards* or *functions* return HTTP 200 and 0 bytes.

So off master, carbonapi and carbonzipper are broken in two different ways, neither of which allows it work with a graphite-web instance.

Does anyone have the stack working with graphite-web 1.1.x and if so, can you point me to a SHA/version release that works and maybe some sample configs?

**Comment by Civil** [comment]
what's the response carbonzipper returns with format=json?

Your symptoms sounds like carbonzipper always returns empty response (empty response is known to be 0 bytes as of now).

Also what do the carbonzipper have in it's logs while doing the queries?

**Comment by reyjrar** [comment]
carbonzipper returns correctly for the `/metrics/` API paths and for any query without functions or wilcards via graphite-web.

carbonapi handles everything except '/metrics/...?format=pickle`, these are always HTTP 200 0 bytes.  If I change that to `format=json`, 100% **correct** behavior.  However, graphite-web uses `format=pickle` for those requests.

**Comment by deniszh** [comment]
@reyjrar 
> If I change that to format=json, 100% correct behavior. However, graphite-web uses format=pickle for those requests.

That's tunable, starting from graphite-web 1.1.1. You can use URL with parameters e.g.
`CLUSTER_SERVERS=['127.0.0.1', 'http://192.168.8.8/graphite?format=json&local=0']`

**Comment by deniszh** [comment]
@reyjrar : Ah, probably I'm wrong - not sure if json will work here out of the box. I see `pickle` and `msgpack` in implementation - https://github.com/graphite-project/graphite-web/pull/2123

**Comment by deniszh** [comment]
Nah, sorry for misleading - it still supporting msgpack or pickle only - https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/finders/remote.py#L305-L321
Then it needs to be fixed in carbonapi side...

**Comment by reyjrar** [comment]
FWIW: `format=msgpack` with carbonapi also produces an HTTP 200, 0 byte response

**Comment by reyjrar** [comment]
@deniszh is maybe the easiest path adding support for format=json in the remote finders for graphite-web?  Would that be a reasonable patch to submit?

**Comment by Civil** [comment]
@reyjrar I cannot reproduce your issue with carbonzipper and wildcards:
```
$ wget -q -O- "http://localhost:8082/metrics/find/?query=carbon.*.felixng.*&format=pickle" | wc -c
2118
```
```
$ wget -q -O- "http://localhost:8082/metrics/find/?format=pickle" | wc -c
211
```

Zipper's config I've used:
```
listen: ":8082"
maxProcs: 0
graphite:
    host: "localhost:2003"
    interval: "60s"
    prefix: "carbon.zipper"
    pattern: "{prefix}.{fqdn}"
buckets: 10
timeouts:
    render: "10s"
    find: "2s"
    connect: "200ms"
concurrencyLimit: 0
keepAliveInterval: "30s"
maxIdleConnsPerHost: 100
expireDelaySec: 10
backendsv2:
  backends:
    -
        groupName: "go-carbon"
        protocol: "auto"
        lbMethod: "broadcast"
        servers:
            - "http://127.0.0.1:8080"
graphite09compat: false
logger:
   -
       logger: ""
       file: "stdout"
       level: "debug"
       encoding: "json"
       encodingTime: "iso8601"
       encodingDuration: "seconds"
```

However I can indeed reproduce that if you'll use carbonapi and not carbonzipper.

With carbonapi I indeed found an issue, for `/metrics/find/?query=...&format=pickle` it always returns empty result, no matter what's the query is. I've pushed a fix for that

As about functions - currently all tag-related functions is supported only if the backend is https://github.com/lomik/graphite-clickhouse . For all other cases graphite-web do not pass the function to CLUSTER_SERVER (at least as far as I remember).

go-carbon's carbonserver (component that's used by carbonapi/carbonzipper with go-carbon setup) currently cannot handle requests that have functions there, but that's an issue with carbonserver.

**Comment by deniszh** [comment]
@reyjrar 
> @deniszh is maybe the easiest path adding support for format=json in the remote finders for graphite-web? Would that be a reasonable patch to submit?

Yes, could be. I remember that someone wanted to create that patch because pickle and msgpack performance issues, but at the end solved those issues differently (e.g. installing different msgpack module).

**Comment by reyjrar** [comment]
@Civil that patch did the trick for carbonapi! Running good now!  I'll be starting to migrate production traffic to it! Thanks!

**Comment by aptituz** [comment]
I did some tests, backporting the fix to the 0.11 branch, and I think I can confirm additionally that the patch does trick as @reyjrar said.

**Comment by Civil** [comment]
Should be fixed in current 0.12.0-rc.0 and in 0.11.1
